### ISSUE TITLE
HLS Live stream targetDurationUs is not always a ideal way to calculate time for next playlist load

### ISSUE THREAD

**Issue by fgl27** [google/ExoPlayer#6360]
### [REQUIRED] Content description
This issue is related to https://github.com/google/ExoPlayer/issues/4217

I also have some content that result in the main issue described here https://github.com/google/ExoPlayer/issues/4217#issue-320953832 

> The issue i'm seeing is when I'm on a good connection, I often get into a very short buffering states (around a second) after playing the stream for a few seconds.

stead of just hardcode the value for **targetDurationUs** as is described there I did this:
https://github.com/fgl27/ExoPlayer/commit/8f298bb23fc7fc061702dd853187ef4d86dfcb9f

But simply setting targetDurationUs to 1s also fix the issue and doesn't seems to add any other, but may cause unnecessary playlist loads.

Including the infromation from https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#page-21
#EXT-X-TARGETDURATION says that all Media Segments will be less or equal to defined value.

this means that we can't restrict the load of a new playlist base on this value (or we can but a furter code check needs to be done to prevent the issue)
as we may have very big #EXT-X-TARGETDURATION = 10 and a small TS size #EXTINF = 1

**this issue usually happens at the beginning of a new playback, after the player has be playing for 30+ seconds all is fine and this doesn't occurs anymore for some time**

I know my change is probably not ideal as is a live stream only issue and if the value of the lowest **#EXTINF** will be used, there is probably a better way to get the value, as if this was supposed to be used on a m3u8 list that is very long will take a lot of time to get the proper value, and the issue may be easier to fix in somewhere else.

I assume if I get this issue others will also get and a proper fix must be done, took me a long time to find that old issue and figure out that I had the same problem.

### [REQUIRED] Link to test content

The content is live and can only be accessed on that moment the link was created, any link shared here will expire or not work as the stream went offline, but I do have the info from the m3u8 files

	#EXT-X-TARGETDURATION:6
	#EXTINF:4.166,live
	#EXTINF:2.000,live
	#EXTINF:1.084,live

### [REQUIRED] Version of ExoPlayer being used
2.10.4

### [REQUIRED] Device(s) and version(s) of Android being used
TV devices running Android 8+

### [REQUIRED] A full bug report captured from the device
Not really device specific issue or something related to logs. This devices that I have don't realy provide usable logs.

**Comment by fgl27** [comment]
Using player.getTotalBufferedDuration() I manage to understand how this issue is be produce.

we are playing a live stream that has as base this values

	#EXT-X-TARGETDURATION:6
	#EXTINF:4.166,live
	#EXTINF:2.000,live
	#EXTINF:1.084,live

the rule for **#EXT-X-TARGETDURATION** says that all Media Segments will be less or equal to defined value so is not a problem that the TARGETDURATION = 6 and EXTINF lower then that;

The value for **TARGETDURATION** never changes, the playback starts the value for **player.getTotalBufferedDuration()** stays at 6 to 7s for about 10s of playback (setBufferDurationsMs is allowing the value to go above this it can't because is live content), after this initial 10s the **player.getTotalBufferedDuration()** stars to get lower, lower than 6s,
[loadPlaylist()](https://github.com/google/ExoPlayer/blob/dev-v2/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/playlist/DefaultHlsPlaylistTracker.java#L492) is triggered but the segments loaded aren't enough to fill the 6s duration of TARGETDURATION, the **player.getTotalBufferedDuration()** get higher but is less than 6s, the buffers keeps going down and down until it reaches 0, at this point **loadPlaylist()** already triggered again but is too late Player.STATE_BUFFERING already trigger the playback has stopped and when the segments are available again the **player.getTotalBufferedDuration()** may go above 6s or not and stay like this for some time, until eventually goes above 6s stays like that for some time but eventually will drop and the issue shows up again and again, can take minutes to happen sometimes but it does all the time.

Keep the value **targetDurationUs = 1s** prevent this from ever happen, but causes unnecessary loadPlaylist() calls every 1s in the case TS are only be created at a 2s rate for example.

So the scenario that probably can occurs to any live stream where at some point has a amount of segments with they added duration time is almost equal or less than targetDurationUs we have this issue.

The first thing that comes to mind is not only call **loadPlaylist()** on the situation that **targetDurationUs** demands but also on the situation that the buffer has already consume all the segments available, not constantly all the time to prevent a race, but at the time the segments are all consumed, and then again in 1s or so if the server didn't provide enough on last update call.

Take a look here and see what can be done, I think I provide enough info about the issue, if need test or more info let me know.

**Comment by marcbaechinger** [comment]
@AquilesCanta Can you please comment on this issue?

**Comment by ojw28** [comment]
> this means that we can't restrict the load of a new playlist base on this value

The HLS specification explicitly says that clients must do this, at least under some circumstances:

> If the client reloads a Playlist file and finds that it has not changed then it MUST wait for a period of one-half the target duration before retrying.

To me, this implies that segment durations probably shouldn't be less than one-half of the target duration either, where-as in the snippet above it seems that segment durations are much shorter.

**Comment by fgl27** [comment]
> To me, this implies that segment durations probably shouldn't be less than one-half of the target duration either, where-as in the snippet above it seems that segment durations are much shorter.

If  the below is a rule
> Media Segments will be less or equal to EXT-X-TARGETDURATION

exo should account for it and not expected they be bigger then half.

this:
https://github.com/google/ExoPlayer/blob/dev-v2/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/playlist/DefaultHlsPlaylistTracker.java#L657

is never be used on this issue I have checked is not a problem of empty/unchanged playlist.

The issue is that exo doesn't account that it can load segments that will not fill up the **targetDurationUs** and base the next playlist load only on that.

I have be playing this same playlist for years now (on non android devices) and never had a issue on any player, I only start to see it when using exoplayer, so others players account for this as is expected...

They do the time verification for next playlist call based on the duration of the already visible segments without expecting those be equal to EXT-X-TARGETDURATION or at least some relation of it.

**Comment by ojw28** [comment]
I'm not sure I understand your response. The HLS specification states:

> When a client loads a Playlist file for the first time or reloads a
> Playlist file and finds that it has changed since the last time it
> was loaded, the client MUST wait for at least the target duration
> before attempting to reload the Playlist file again, measured from
> the last time the client began loading the Playlist file.
>
> If the client reloads a Playlist file and finds that it has not
> changed, then it MUST wait for a period of one-half the target
> duration before retrying.

The code you reference appears to do exactly that, in which case we're implementing the specification correctly. Conversely, you appear to be saying that we should implement something that's explicitly not what the HLS specification asks for. Please clarify.

**Comment by fgl27** [comment]
from where you are reading those specification? share the link please.

tell me if this is also a " HLS specification"

> Media Segments will be less or equal to EXT-X-TARGETDURATION

if it is and is be used here so I guess yes, there is the chance that you load a playlist that has on it segments that they total time is less then **EXT-X-TARGETDURATION**, in that case the next playlist can't be load base on **EXT-X-TARGETDURATION** alone, because if that happens this issue I open will happens.

So if "Media Segments will be less or equal to EXT-X-TARGETDURATION" is a rule, the way exo loads playlist can't be used as it directly conflict on how the total time of this new playlist can be.

so does exoplayer account for the total time of the segments loaded be less then next TARGETDURATION?
NO

can it do?
YES

So do you understand the issue now?

can it be fixed and all the rules even those that conflict be addressed?

**Comment by ojw28** [comment]
> from where you are reading those specification? share the link please.

The same specification that you've already linked to. [Section 6.3.4](https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-6.3.4).

> the next playlist can't be load base on EXT-X-TARGETDURATION alone

The specification is very clear about the minimum delay between playlist refreshes, and your proposal explicitly goes below them. It's unlikely that this is correct.

> there is the chance that you load a playlist that has on it segments that they total time is less then EXT-X-TARGETDURATION, in that case the next playlist can't be load base on EXT-X-TARGETDURATION alone, because if that happens this issue I open will happens.

On further thought, I don't think this is correct. Provided the playback position isn't too close to the live edge, the segments can be arbitrarily short and playback should work fine with the current playlist refresh behaviour. You'll just see more than one new segment each time the playlist is refreshed, which is fine. Are you doing anything special to get the playback position closer to the live edge than it is by default? Is the initial playback position closer to the live edge than it is for other players?

**Comment by fgl27** [comment]
No I'm not, I notice that the playback is as close to the edge as can be, the only thing I have change is that the default buffers are lower DEFAULT_BUFFER_FOR_PLAYBACK_MS and DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS are 1s.

When the **HlsMediaPlaylist** is called the total segment size is 28s, but it isn't starting on the first segment but 6s before the end.

With is odd as that is the **targetDurationUs** size, here is very noticeable, I start the playback player.getTotalBufferedDuration() goes to 6 and starts to drop, reaches 0, loadPlaylist() is called loades more 6s and stay like this, the problem is that it does not do fast enough so **Player.STATE_BUFFERING** is always triggered and playback stops every 6s after the issue starts.

What in exoplayer could cause this behavior?

Exo see that there is 28s of segments but start at the last 6s (targetDurationUs = 6), and because of that the issue is display.

The player other than the default buffer sizes changes, I start with

	player.prepare(mediaSource, true, true);

No seek or anything related before that is called for live content...

**Comment by AquilesCanta** [comment]
I think there are two different topics in discussion here:
- Quick rebuffer after starting playback.
- Playlist reloading policy.

On the quick rebuffer topic:

> playback starts 6s to the live-edge.

ExoPlayer's default start position for live playback is 3 segments from the live edge. In your case, your last 3 segments' total duration is 6 seconds. The target duration is 6 seconds, so we won't reload the playlist for 6 seconds, time after which the buffer will surely be exhausted. **Note**: After some rebuffering, playback should stabilize unless the window is 3 segments long (please don't do that, may be worth reading [this post](https://medium.com/google-exoplayer/hls-playback-in-exoplayer-a33959a47be7)).

There are two ways of fixing this, as far as I can see:
- You can set an [EXT-X-START](https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-4.3.5.2) of -12 (means playback will start 12 seconds from the live edge).
- You can implement your own playlist tracker that reloads the media playlist as you see fit.

On the playlist reloading policy: 
The standard is pretty clear about how to reload media playlists. I doubt we should adjust it for specific cases, since we might create issues for others (for example, by DOSing).

By the way, nice chimarrão.

**Comment by ojw28** [comment]
Agreed the policy about playlist reloading is clear. Given this, we should probably change our default start position to better handle cases where segments are significantly shorter than the target duration. For example:

1. Start at least 3 segments back from the live edge (this is the current rule)
1. Also start at least 2 (or 3?) target durations back from the live edge (this would be an additional rule to handle cases such as the one in this issue)

What do you think?

**Comment by fgl27** [comment]
So I check and most stream I tested before the TS was 2s so the 6s issue comes from this edge position and target duration be equal, as when was a bigger TS I didn't notice the issue, but those TS size are rare.

> * Start at least 3 segments back from the live edge (this is the current rule)
> * Also start at least 2 (or 3?) target durations back from the live edge (this would be an additional rule to handle cases such as the one in this issue)

This seems OK, maybe a simple check **if (3 segments size <= target durations) starting point = 3 segments size + target durations;**

Also check if there is that much segments in time if not start on the first one (maybe the code already does that).
I know you guys know how to deal with that logic, just a idea.

**Comment by AquilesCanta** [comment]
We can change the default start duration to start from 

```
Math.min(
    segments.isEmpty()
        ? 0
        : segments.get(Math.max(0, segments.size() - 3)).relativeStartTimeUs,
    playlist.durationUs - playlist.targetDurationUs * 2);
```
Using 2 will keep the existing behavior for most users while reducing the risk of entering buffering state. If we want to use 3 * target duration we can just remove the old behavior.

Note @fgl27 that you should set the default start position in the playlist as suggested above . That way you can guarantee a consistent playback experience across all platforms.

**Comment by fgl27** [comment]
That worked, **player.getTotalBufferedDuration()** is now oscillating from 5 to 12s, no rebuffers call, I can see the playlist update it 6s as expected, all is just fine.

:+1:

**Comment by AquilesCanta** [comment]
Please clarify what specifically you tried for future reference.

**Comment by fgl27** [comment]
> Please clarify what specifically you tried for future reference.

This https://github.com/fgl27/ExoPlayer/commit/698078fd23e36be43e99ac3805c606b48799ca67

targetDurationUs = 6s (default value from this particularly m3u8 list)
segments = 2s

Before the player started 6s from the live edge as segments was 2s we get the issue, now after https://github.com/fgl27/ExoPlayer/commit/698078fd23e36be43e99ac3805c606b48799ca67 starts at 12s and the issue described on post 1 and 2 no longer happens.

**Comment by AquilesCanta** [comment]
Ok, a fix in the library will be available in the dev-branch soon. It may be slightly better because it aligns the start position to the start of a segment.

**Comment by fgl27** [comment]
OK I check that when available.

**Comment by fgl27** [comment]
@ojw28 and @AquilesCanta 

I just want to let you guys know that even tho this commit https://github.com/google/ExoPlayer/commit/0aba89b60e02e4b0aab9ad518d71d85d71480491 hepls doesnot prevent the issue from happening.

why?

the title already says it all **HLS Live stream **targetDurationUs** is not always a ideal way to calculate time for next playlist load**

I know you guys may be thinking this guy will never stop whit that :stuck_out_tongue_closed_eyes:, no in this cause because the problem isn't totally solved.

if the bellow keep be a rule, and keeps be used the issue will prevail

> #EXT-X-TARGETDURATION says that all Media Segments will be less or equal to defined value.

and we know that it is be used not just in my case but on the old issue that had similar problem and probably will happen again for someone in the future.

Yes I use https://github.com/google/ExoPlayer/commit/0aba89b60e02e4b0aab9ad518d71d85d71480491  and yes the isseu keeps coming up eventually very dealyded but it comes.

The reason is because Exoplayer doesn't deal with the issue that a playlist may load, new segments will be available but the time of the already visible segments may be lower than the next playlist load, that is because yes the segments be used are lower then **targetDurationUs** but that is allowed.

Other because the we have a "lag" on the server (even tho targetDurationUs has be passed the server didn't manage to include enough segmentes to fill up the time until the next targetDurationUs aka next playlist load) or client side this issue will keep happening for streams that follow the rule to the letter and use segments sizes lower then **targetDurationUs**.

So do you guys plan to address the issue?

This is not a problem for me as I can do this https://github.com/fgl27/ExoPlayer/commit/ab2d94931a77839c58402a1a78964f88efc51ed0 

But I just care that the issue will happen to someone  and will probably take a long time to understand why and may never find a solution for the random "hiccups" caused by Player.STATE_BUFFERING call when there is no more segments to load...

**Comment by fgl27** [comment]
Just in relation to the question of https://github.com/google/ExoPlayer/issues/4217#issue-320953832 **Question on achieving a low latency HLS live stream**

Exo doesn't provide a way to get as close as possible of live edge when the playlist doesn't included  **EXT-X-START**, after reading the above issue and address this issue I keep thinking on how to do it without using seekto (because is unreliable as the duration size may be unknown), after all this streams I'm playing are somewhat low latency ready that is way they have small TS sizes but they don't provide **EXT-X-START**.

So I made this https://github.com/fgl27/ExoPlayer/commit/f379291b668a10cf27d921b5dde638ebe414aec1

I know is a workaround, but it does allows to control how close we get from live edge and allows us to override #EXT-X-TARGETDURATION **targetDurationUs** and choose  from the app side when we generate MediaSource how close we will be from the edge, as I already build the player from source this is easier then override and implement a in app **HlsMediaSource.Factory** or something related.

with this change and a lower value **DEFAULT_BUFFER_FOR_PLAYBACK_MS** (lower then LowLatency) and a **DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS** equal or lower to the LowLatency value we can achieve a Low Latency scenario

I tested this and seems fairly stable, some times **DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS** is triggered but only once as long as we keep it as equal as the LowLatency value.

Of course for this is needed a good internet connection and TS that are small.

just in case someone on the future research on how to achieve low latency on live stream and exoplayer this is a OK workaround that I will if necessary improve by me, with luck the person will find this on they search.

**Comment by tonihei** [comment]
@fgl27 We already track support for low-latency HLS in #5011. The work in this area will address the issue you mention in some ways because we likely need to update our playlist load strategy again.

**Comment by ojw28** [comment]
> The title already says it all HLS Live stream targetDurationUs is not always a ideal way to calculate time for next playlist load

The HLS specification is very unambiguous about requiring that clients wait a minimum period of time between playlist updates. Repeatedly saying the same thing does nothing to change that.

> The reason is because Exoplayer doesn't deal with the issue that a playlist may load, new segments will be available but the time of the already visible segments may be lower than the next playlist load, that is because yes the segments be used are lower then targetDurationUs but that is allowed.

We understand and agree that segments may be less than the target duration. The logic we're using handles this case. As an example, consider an extreme case where target duration is 10s and each segment is much shorter, let's say 1s. Using the new logic, we'll start playback 20s behind the live edge. At this point, we have 20s of media that we can play without refreshing the playlist, so refreshing it after the target duration of 10s is going to be fine. When we do this we should see 10 new 1s segments added to the playlist. We'll once again have 20s of media that we can play without refreshing the playlist, and so we can wait another 10s, and so on. Note that the shortness of the segments is completely irrelevant to this working correctly. If the segments were 0.5s refreshing at the target duration is **still** fine. The only difference is that in each playlist update, we'll see 20 new 0.5s segments rather than 10 new 1s segments.

> Other because the we have a "lag" on the server (even tho targetDurationUs has be passed the server didn't manage to include enough segmentes to fill up the time until the next targetDurationUs aka next playlist load) or client side this issue will keep happening for streams that follow the rule to the letter and use segments sizes lower then targetDurationUs.

I don't really understand this, but it's the servers responsibility to make new segments available in a predictable and timely manner.

**Comment by fgl27** [comment]
All I saying is that I have tested yours change and it eventually allows the issue to happens.

Yes the playback started two **targetDurationUs** away from the edge, but after some time some minutes 10 to 30 I notice that the **player.getTotalBufferedDuration()**  was getting lower and lower (isn't always but at some point in some streams it does), I don't know way exactly (I wrote those post above in assumptions of what is probably happening), and then eventually the issue started all over again because the **player.getTotalBufferedDuration()** is now equal or lower then **targetDurationUs**.

When I hardcoded the **targetDurationUs** to 1s I have zero problems like this stream plays for as long as I wanna no rebuffering ever, even using my workaround of starting playing 3s from the edge.

So all I wanna is for you guys be aware that the solution applied here is not enough, I would not be bothering with the issue if it had solved the problem, it does help but doesn't solves it.
